accounting/src/delta/delta_data_collection/mod.rs

```diff
@@ -110,170 +140,111 @@ impl<K: Ord + Copy, T> DeltaDataCollection<K, T> {
+        DeltaDataCollection {
+            data: BTreeMap::<K, DeltaMapElement<T>>::from_iter(
+                iter.into_iter().map(|(k, d)| (k, DeltaMapElement::Delta(d))),
+            ),
         }
     }
 }
 
-impl<K: Ord + Copy, T> FromIterator<(K, DataDelta<T>)> for DeltaDataCollection<K, T> {
-    fn from_iter<I: IntoIterator<Item = (K, DataDelta<T>)>>(iter: I) -> Self {
-        DeltaDataCollection {
-            data: BTreeMap::<K, DataDelta<T>>::from_iter(iter),
-        }
+/// This function returns a delta that if applied to the result of merge(delta1,delta2) gives original delta1
+fn create_undo_delta<T: Clone>(
+    lhs: Option<&DataDelta<T>>,
+    rhs: DataDelta<T>,
+) -> Result<DataDelta<T>, Error> {
+    match lhs {
+        Some(lhs) => match (lhs, rhs) {
+            (DataDelta::Create(d1), DataDelta::Create(d2)) => {
+                Ok(DataDelta::Modify((d2, d1.clone())))
+            }
+            (DataDelta::Create(d1), DataDelta::Modify((_, d2))) => {
+                Ok(DataDelta::Modify((d2, d1.clone())))
+            }
+            (DataDelta::Create(_), DataDelta::Delete(d)) => Ok(DataDelta::Create(d)),
+
+            (DataDelta::Modify((_, d1)), DataDelta::Create(d2)) => {
+                Ok(DataDelta::Modify((d2, d1.clone())))
+            }
+            (DataDelta::Modify((_, prev)), DataDelta::Modify((_, new))) => {
+                Ok(DataDelta::Modify((new, prev.clone())))
+            }
+            (DataDelta::Modify((_, _)), DataDelta::Delete(d)) => Ok(DataDelta::Create(d)),
+
+            (DataDelta::Delete(_), DataDelta::Create(d)) => Ok(DataDelta::Delete(d)),
+            (DataDelta::Delete(_), DataDelta::Modify((_, _))) => {
+                Err(Error::DeltaDataModifyAfterDelete)
+            }
+            (DataDelta::Delete(_), DataDelta::Delete(_)) => {
+                Err(Error::DeltaDataDeletedMultipleTimes)
+            }
+        },
+        None => match rhs {
+            DataDelta::Create(d) => Ok(DataDelta::Delete(d)),
+            DataDelta::Modify((prev, new)) => Ok(DataDelta::Modify((new, prev))),
+            DataDelta::Delete(d) => Ok(DataDelta::Create(d)),
+        },
     }
 }
 
 /// Given two deltas, combine them into one delta, this is the basic delta data composability function
-fn combine_delta_data<T>(
+fn combine_delta_data<T: Clone>(
     lhs: &DataDelta<T>,
     rhs: DataDelta<T>,
-) -> Result<DeltaMapOp<DataDelta<T>>, Error> {
+) -> Result<DataDelta<T>, Error> {
     match (lhs, rhs) {
-        (DataDelta::Create(_), DataDelta::Create(_)) => Err(Error::DeltaDataCreatedMultipleTimes),
-        (DataDelta::Create(_), DataDelta::Modify(d)) => Ok(DeltaMapOp::Write(DataDelta::Create(d))),
-        (DataDelta::Create(_), DataDelta::Delete) => {
-            // if lhs had a creation, and we delete, this means nothing is left and there's a net zero to return
-            Ok(DeltaMapOp::Delete)
+        (DataDelta::Create(_), DataDelta::Create(d)) => Ok(DataDelta::Create(d)),
```

Comment:
Why are we now accepting two consecutive Create ops? This should be an error.

Reply:
To be clear, it was an error... why was this function changed this way?

Reply:
This is valid case if we undo, because undo for Delete is Create.
Take a look at `create_delete_undo_merge` test.

Reply:
Though I agree that it's confusing. Maybe we need some kind of `UndoDelta` that would identify undo operation and not user operation

Reply:
We definitely need that... this is absolutely unreadable and even incorrect. In its current form, if a user creates twice, it'll succeed at the lowest level, and high levels should detect the error. This isn't acceptable.